### DISCUSSION TITLE
How to receive steam messages

### DISCUSSION THREAD

**Discussion by Detla7t** [ValvePython/steam#391] [Q&A] [answered]
hi so i've been using this library for the past few days and i just can figure out how to get a message from another user or a group ive looked at https://github.com/ValvePython/steam/issues/13 as well as the docs and i found some things that looked correct but this i think is the closest ive gotten but it still doesn't work 
`
client.cli_login()
while True:
    @client.on(EMsg.ClientChatMsg)
        def handle_message(user: SteamClient, message_text):
            user.user.send_message('You said: %s' % message_text)
            print("Message Handled?")`
ive tried the EMsg ClientChatMsg, ClientFriendMsg, ClientFriendMsgIncoming, ClientFriendMsgEchoToSender, ClientChatOfflineMessageNotification, ClientChatRequestOfflineMessageCount, ClientChatGetFriendMessageHistory, ClientChatGetFriendMessageHistoryResponse, ClientChatGetFriendMessageHistoryForOfflineMessages, UMQSendChatMessage, and UMQIncomingChatMessage but with no results i am aswell logging in before hand and even looping over the thing but i dont know why i cant get a steam message or even the print i setup to say "Message Handled?" can some one help me please.

**Comment by Gobot1234** [comment, marked as answer]
To handle messages you need to listen for the `"chat_message"` event you also shouldn't register the listener inside of a while True loop, your code should like more like
```py
client = ...

@client.on("chat_message")
def on_message(user: steam.client.user.User, text: str):
    ...

client.cli_logon()
```
Recieving messages from a group/clan currently isn't handled by the client by default see #13 for more. You can implement for them yourself with a bit of fandangling.
```py
@client.on("ChatRoomClient.NotifyIncomingChatMessage#1")
def on_clan_or_group_message(msg):
    ... # go wild with msg, (it should be a wrapper around https://github.com/ValvePython/steam/blob/bf3cb330d3e4303a20092e842c503e373294382f/protobufs/steammessages_chat.proto#L786-L797)
```

**Reply by Detla7t** [reply]
hi so following your examples sorry i still dont understand what im not doing correct here is all my test code.

        import steam.client.user 
        import steam.client
        import json
        
        client = steam.client.SteamClient()
        LoginData = json.loads(open("login.json", "r").read())
        
        @client.on(event="chat_message")
        def on_message(user: steam.client.user.SteamUser, text: str):
            print("ReceivedMessage:",text)
        
        client.cli_login(username=LoginData["Username"],password=LoginData["Password"])
        
        stop = ""
        
        while stop != "stop":
            stop = input("type stop to Stop:")
when i send a message from another account it just doesn't show them in the console i dont know if its because the while loop is blocking execution or what and if so how should i go about stopping it from blocking execution? or am i doing something else wrong? thanks so much for the reply :)

**Reply by Gobot1234** [reply]
input will block gevents event loop, I'd recommend using client.run_forever() and then just close the program using `ctrl + c`

**Reply by Detla7t** [reply]
thanks so much again the final code i have looks like this and works

    import steam.client.user 
    import steam.client
    import json
    
    client = steam.client.SteamClient()
    LoginData = json.loads(open("login.json", "r").read())
    
    @client.on(event="chat_message")
    def on_message(user: steam.client.user.SteamUser, text: str):
        print("ReceivedMessage:",text," From:",user)
    
    client.cli_login(username=LoginData["Username"],password=LoginData["Password"])
    client.run_forever()

and i got this output `ReceivedMessage: test  From: <SteamUser(steamid, EFriendRelationship.Friend, EPersonaState.Online)>` thank you so much :)
